torch/csrc/distributed/c10d/TraceUtils.h

```diff
@@ -475,16 +492,61 @@ struct NCCLTraceBuffer {
     return result;
   }
 
-  void retire_id(c10::optional<size_t> id) {
+  /*
+  Mark an Event as completed and free its events.
+
+  This is called by the watchdog thread, and is asynchronous from the
+  perspective of the main thread.
+
+  compute_duration defaults to true since retire_id is only called in the
+  watchdog thread, which is currently a place we call cuda APIs which may hang,
+  but care should be taken to avoid computing duration in any function that must
+  never hang. (timing must also be enabled for compute_duration - see
+  TORCH_NCCL_ENABLE_TIMING).
+  */
+  void retire_id(c10::optional<size_t> id, bool compute_duration = true) {
     if (!enabled_ || !id) {
       return;
     }
+
+    bool can_compute_duration = false;
+    EventList* startEvents = nullptr;
+    EventList* endEvents = nullptr;
+    c10::optional<float> duration = c10::nullopt;
+
     std::lock_guard<std::mutex> guard(mutex_);
+
     auto& entry = entries_.at(*id % max_entries_);
     if (entry.id_ == *id) {
       update_state(entry);
+
+      if (compute_duration) {
+        can_compute_duration = strcmp(entry.state_, "completed") == 0 &&
+            entry.start_ && entry.end_;
+        startEvents = entry.start_;
+        endEvents = entry.end_;
+      }
+    }
+
+    mutex_.unlock();
+    // Compute duration without without holding the lock, because
+    // cudaEventDuration() can hang, and we need to acquire the lock before we
+    // can dump(), which we never want to block.
+    if (can_compute_duration) {
+      duration = getDurationFromFirstEvent(*startEvents, *endEvents);
+    }
+    mutex_.lock();
+
+    if (entry.id_ == *id) {
+      if (duration.has_value()) {
+        entry.duration_ = duration.value();
+      }
       entry.retired_ = true;
       entry.start_ = entry.end_ = nullptr;
+    } else {
+      LOG(INFO)
+          << "Failed retire_id " << *id
```

Comment:
I wouldn't use the word Failed here. If we hit this, it just means that it was configured to no longer be interested in this event.

Reply:
point taken. user doesn't care about this event anymore. (but its kind of unexpected / buggy (?) if we take so long to reacquire the lock that our whole buffer gets overwritten before we can compute durations and mark finished) -- worst case this could show up as durations are never working or are spotty in the part of the flight recorder you do dump